src/ray/object_manager/plasma/create_request_queue.cc

```diff
@@ -24,31 +24,148 @@
+    RAY_LOG(ERROR)
+        << "Object store client requested the result of a previous request to create an "
+           "object, but the result has already been returned to the client. This client "
+           "may hang because the creation request cannot be fulfilled.";
+    return false;
+  }
+
+  if (!it->second) {
+    return false;
+  }
+
+  *result = it->second->result;
+  *error = it->second->error;
+  fulfilled_requests_.erase(it);
+  return true;
+}
+
+std::pair<PlasmaObject, PlasmaError> CreateRequestQueue::TryRequestImmediately(
+    const ObjectID &object_id, const std::shared_ptr<ClientInterface> &client,
+    const CreateObjectCallback &create_callback) {
+  PlasmaObject result = {};
+
+  if (!queue_.empty()) {
+    // There are other requests queued. Return an out-of-memory error
+    // immediately because this request cannot be served.
+    return {result, PlasmaError::OutOfMemory};
+  }
+
+  auto req_id = AddRequest(object_id, client, create_callback);
+  if (!ProcessRequests().ok()) {
+    // If the request was not immediately fulfillable, finish it.
+    RAY_CHECK(!queue_.empty());
+    FinishRequest(queue_.begin());
+  }
+  PlasmaError error;
+  RAY_CHECK(GetRequestResult(req_id, &result, &error));
+  return {result, error};
+}
+
+Status CreateRequestQueue::ProcessRequest(std::unique_ptr<CreateRequest> &request) {
+  // Return an OOM error to the client if we have hit the maximum number of
+  // retries.
+  bool evict_if_full = evict_if_full_;
+  if (max_retries_ == 0) {
+    // If we cannot retry, then always evict on the first attempt.
+    evict_if_full = true;
+  } else if (num_retries_ > 0) {
+    // Always try to evict after the first attempt.
+    evict_if_full = true;
+  }
+
+  request->error = request->create_callback(evict_if_full, &request->result);
+  Status status;
+  auto should_retry_on_oom = max_retries_ == -1 || num_retries_ < max_retries_;
+  if (request->error == PlasmaError::TransientOutOfMemory) {
+    // The object store is full, but we should wait for space to be made
+    // through spilling, so do nothing. The caller must guarantee that
+    // ProcessRequests is called again so that we can try this request again.
+    // NOTE(swang): There could be other requests behind this one that are
+    // actually serviceable. This may be inefficient, but eventually this
+    // request will get served and unblock the following requests, once
+    // enough objects have been spilled.
+    // TODO(swang): Ask the raylet to spill enough space for multiple requests
+    // at once, instead of just the head of the queue.
+    num_retries_ = 0;
+    status =
+        Status::TransientObjectStoreFull("Object store full, queueing creation request");
+  } else if (request->error == PlasmaError::OutOfMemory && should_retry_on_oom) {
+    num_retries_++;
+    RAY_LOG(DEBUG) << "Not enough memory to create the object, after " << num_retries_
+                   << " tries";
+
+    if (on_store_full_) {
```

Comment:
It's probably more explicit to call this trigger_global_gc instead of on_store_full, so the reader doesn't need to wonder what this is triggering.